File: Vegas_Integration/plot_macro.py
## @file plot_macro.py
# Python code to generate a plot from the
# VegasHarmOsc.C output.
#
# Python code to generate a plot from the
# VegasHarmOsc.C output. This macro also performs
# a fit to the function fitplot_py(), providing an
# estimate for the ground energy E_0. Various graphical
# options are implemented and tuned in the code. Type:\n
# $ python plot_macro.py\n
# to run and generate the plot "plot_py.png".

import re
import numpy as np
import matplotlib.pyplot as plt
from numpy import genfromtxt
from scipy.optimize import least_squares

## Function to read time_bound from header file
#
#   @param filename name of the file to look into.
#   @see SETTINGS.h
def read_param_from_file(filename):
    regexp = re.compile(r' *double time_bound *(=) *([0-9.-]+); *.*')
    time_bound=0.
    with open(filename) as f:
        for line in f:
            match = regexp.match(line)
            if match:
                print("Found in header file time_bound =", match.group(2))
                time_bound=match.group(2)
    return time_bound

## Fit function to be used in the scipy.optimize.least_squares method
#
#   @param par array of parameters, where par[0]=E_0 and par[1]=T
#   @param x coordinate x
#   @param y data to be fitted
#   @see physical_params
def fitfunction_py(par,x,y):
    return (np.exp(-x**2-par[0]*par[1])/np.sqrt(3.14159))-y

## Fit function to be used in the plot with fitted paramaters
#
#   @param x coordinate x
#   @param E_0 ground state energy
#   @param T propagation time
#   @return \f[\left|\left\langle \psi | E_{0}\right\rangle\right|^{2}e^{-E_{0}T}=\frac{1}{\sqrt{\pi}}e^{-x^{2}-E_{0}T}\f]
#   @see physical_params
def fitplot_py(x,E_0,T):
    return (np.exp(-x**2-E_0*T)/np.sqrt(3.14159))

## \cond MAIN
#************** EXECUTION ******************#

input_name="SETTINGS.h"
output_name="plot_py.png"

data=genfromtxt('output_file.dat', delimiter='\t', skip_header=1)
x=data[:,0]
result=data[:,1]
error=data[:,2]
asymptotic=data[:,3]
n_data=x.size

time_bound=read_param_from_file(input_name)

par=np.zeros(2)
par[0]=0.5
par[1]=time_bound
#Fit data to the fitting function using the least square method implemented in the the scipy function
fit_lsq = least_squares(fitfunction_py, par, bounds=([0.,1.],[time_bound,time_bound]), loss='soft_l1', args=(x[:], result[:]))
x_fit = np.linspace(x[0], x[n_data-1],500) #generation of points for the graph
y_fit = fitplot_py(x_fit, *fit_lsq.x) #generation of points for the graph

print("Generating plot with following data")
print("x     result     error      asymptotic")
for i in range(n_data):
    print(x[i],result[i],error[i],asymptotic[i])

# Plot all data and a legend
plt.errorbar(x,result, yerr=error, color='blue', linestyle="None", marker='o', markersize=4, label="Montecarlo data")
plt.plot(x_fit,y_fit, color='green', marker='None', label="Exact asymptotic for $E_{0}=\frac{1}{2}$")
plt.legend(loc='lower center')
# Set title and x-y axis labels for the Axes
plt.title("1D Harmonic Oscillator with Vegas", fontsize=18)
plt.xlabel('$x$')
plt.ylabel('$<x|e^{-HT}|x>$')
# Adds gridlines to the Axes
plt.grid(linestyle='--', color='gray')
# Save the current figure to a file
plt.savefig(output_name)
plt.show()

## \endcond

File: Vegas_Integration/plot_macro.cpp
///////////////////////////////////////////////////////////
/// \file plot_macro.cpp
/// \brief ROOT macro to generate a plot from the
///        VegasHarmOsc.cpp output.
///
///  ROOT macro to generate a plot from the
///  VegasHarmOsc.cpp output. This macro also performs
///  a fit to the function fitfunction, providing an
///  estimate for the ground energy E_0. Various graphical
///  options are implemented and tuned in the code. Type:\n
///  $ ROOT plot_macro.cpp\n
///  to run and generate the plot "plot_ROOT.png".
///
///////////////////////////////////////////////////////////
#include <algorithm>
#include <cmath>
#include <fstream>
#include <iostream>
#include <string>
#include "TCanvas.h"
#include "TF1.h"
#include "TGraphErrors.h"

///////////////////////////////////////////////////////////
/// Fit function
///
/// Function to fit data to: it represents the exact
/// asymptotic result from standard quantum mechanics
/// \param x pointer to the position x, where x[0]=x
/// \param par parameters of the function: par[0]=E_0, par[1]=T,
///        where E_0 is the ground state energy and T the propagation
///        time
///  \return \f[\left|\left\langle \psi |
///  E_{0}\right\rangle\right|^{2}e^{-E_{0}T}=\frac{1}{\sqrt{\pi}}e^{-x^{2}-E_{0}T}\f] \see
///  physical_params
///
///////////////////////////////////////////////////////////
double fitfunction(double* x, double* par)
{
  return exp(-x[0] * x[0] - par[0] * par[1]) / sqrt(3.14159);
}

int plot_macro()
{
#include "SETTINGS.h"

  // Define the root 1D function to be used in plot and fit
  TF1* f = new TF1("fit", fitfunction, -x_loop_bound, x_loop_bound, 2);

  // Set some useful properties, initialize parameters and fix the propagation time
  // to its tabulated value
  f->SetName("myfit");
  f->SetLineColor(2);
  f->SetParameter(0, 0.5);
  f->FixParameter(1, time_bound);
  f->SetParLimits(0, 0., 1.);
  f->SetParNames("E_0", "T");

  // Initialize a canvas and a pad and set their attributes
  TCanvas* c = new TCanvas("c", "Data and Fit", 800, 800);
  gStyle->SetOptFit(1011);
  gPad->SetGrid(1, 1);
  TPad* mypad = new TPad("mypad", "mypad", 0., 0., 1., 1.);
  mypad->Draw();
  mypad->cd();
  mypad->Range(0, 0, 100, 100);
  mypad->SetLeftMargin(0.15);
  mypad->SetRightMargin(0.15);
  mypad->SetTopMargin(0.15);
  mypad->SetBottomMargin(0.15);

  // Define the object containing the simulated data
  auto montecarlo = new TGraphErrors();

  // Open the data file
  ifstream infile("output_file.dat");
  string waste;
  double x = 0., result = 0., error = 0., exact = 0.;

  // If open, extract all simulated data from the file
  if (infile) {
    infile.ignore(500, '\n');
    for (int i = 0; i < (int)(2 * x_loop_bound / x_loop_step) + 1; i++) {
      infile >> x >> result >> error >> exact;
      montecarlo->AddPoint(x, result);
      montecarlo->SetPointError(i, 0., error);
    }
  } else {
    cout << "ERROR: \"output_file.dat\" not found\n";
    return 1;
  }

  // Tune some graphical aspects
  montecarlo->SetTitle("1D Harmonic Oscillator with Vegas");
  montecarlo->GetXaxis()->SetLimits(-x_loop_bound - 1., x_loop_bound + 1.);
  montecarlo->GetXaxis()->SetRangeUser(-x_loop_bound - 0.5, x_loop_bound + 0.5);
  montecarlo->GetXaxis()->SetLabelSize(0.04);
  montecarlo->GetXaxis()->SetTitleSize(0.04);
  montecarlo->GetXaxis()->SetLabelOffset(0.008);
  montecarlo->GetXaxis()->SetTitleOffset(1.0);
  montecarlo->GetXaxis()->SetTitle("x");
  montecarlo->GetXaxis()->CenterTitle(true);
  montecarlo->GetYaxis()->SetTitle("<x|exp(-HT)|x>");
  montecarlo->GetYaxis()->SetLimits(-1., 1.);
  montecarlo->GetYaxis()->SetRangeUser(0., 1.3 * (f->Eval(0.)));
  montecarlo->GetYaxis()->SetLabelSize(0.04);
  montecarlo->GetYaxis()->SetTitleSize(0.04);
  montecarlo->GetYaxis()->SetLabelOffset(0.012);
  montecarlo->GetYaxis()->SetTitleOffset(2.);
  montecarlo->GetYaxis()->CenterTitle(true);
  montecarlo->SetName("montecarlo");
  montecarlo->SetMarkerColor(4);
  montecarlo->SetMarkerStyle(20);

  // Perform the fit
  auto fitresult = montecarlo->Fit("myfit");
  // Draw the fit as a line and simulated data as points
  montecarlo->Draw("AP");

  // Draw the exact asymptotic function
  f->SetLineColor(3);
  f->SetParameter(0, 0.5);
  f->Draw("lsame");

  // Initialize and draw a legend
  auto legend = new TLegend(0.15, 0.85, 0.45, 0.7);
  legend->AddEntry("myfit", "#splitline{Exact asymptotic}{for E_0=1/2}", "l");
  legend->AddEntry("montecarlo", "Montecarlo data", "ep");
  legend->SetTextSizePixels(18);
  legend->Draw("same");

  // Save plot into a file
  c->Print("plot_ROOT.png");
  return 0;
}
